llvm/lib/Target/DirectX/DXILRootSignature.cpp

```diff
@@ -0,0 +1,224 @@
+    !2 = !{ ptr @main, !3 } ; function, root signature
+    !3 = !{ !4, !5, !6, !7 } ; list of root signature elements
+
+    So for each MDNode inside dx.rootsignatures NamedMDNode
+    (the Root parameter of this function), the parsing process needs
+    to loop through each of its operands and process the function,
+    signature pair.
+ */
+
+  LLVMContext *Ctx = &M.getContext();
+
+  SmallDenseMap<const Function *, ModuleRootSignature> MRSMap;
+
+  NamedMDNode *RootSignatureNode = M.getNamedMetadata("dx.rootsignatures");
+  if (RootSignatureNode == nullptr)
+    return MRSMap;
+
+  for (const auto &RSDefNode : RootSignatureNode->operands()) {
+    if (RSDefNode->getNumOperands() != 2) {
+      reportError(Ctx, "Invalid format for Root Signature Definition. Pairs "
+                       "of function, root signature expected.");
+      continue;
+    }
+
+    // Function was pruned during compilation.
+    const MDOperand &FunctionPointerMdNode = RSDefNode->getOperand(0);
+    if (FunctionPointerMdNode == nullptr) {
+      continue;
+    }
+
+    ValueAsMetadata *VAM =
+        llvm::dyn_cast<ValueAsMetadata>(FunctionPointerMdNode.get());
+    if (VAM == nullptr) {
+      reportError(Ctx, "First element of root signature is not a Value");
+      continue;
+    }
+
+    Function *F = dyn_cast<Function>(VAM->getValue());
+    if (F == nullptr) {
+      reportError(Ctx, "First element of root signature is not a Function");
+      continue;
+    }
+
+    MDNode *RootElementListNode =
+        dyn_cast<MDNode>(RSDefNode->getOperand(1).get());
+
+    if (RootElementListNode == nullptr) {
+      reportError(Ctx, "Missing Root Element List Metadata node.");
+    }
+
+    ModuleRootSignature MRS;
+
+    if (parse(Ctx, MRS, RootElementListNode) || validate(Ctx, MRS)) {
+      return MRSMap;
+    }
+
+    MRSMap.insert(std::make_pair(F, MRS));
+  }
+
+  return MRSMap;
+}
+
+AnalysisKey RootSignatureAnalysis::Key;
+
+SmallDenseMap<const Function *, ModuleRootSignature>
+RootSignatureAnalysis::run(Module &M, ModuleAnalysisManager &AM) {
+  return analyzeModule(M);
+}
+
+//===----------------------------------------------------------------------===//
+
+static void printSpaces(raw_ostream &Stream, unsigned int Count) {
+  for (unsigned int I = 0; I < Count; ++I) {
+    Stream << ' ';
+  }
+}
+
+PreservedAnalyses RootSignatureAnalysisPrinter::run(Module &M,
+                                                    ModuleAnalysisManager &AM) {
+
+  SmallDenseMap<const Function *, ModuleRootSignature> &MRSMap =
+      AM.getResult<RootSignatureAnalysis>(M);
+  OS << "Root Signature Definitions"
+     << "\n";
+  uint8_t Space = 0;
+  for (const auto &P : MRSMap) {
+    const auto &[Function, MRS] = P;
+    OS << "Definition for '" << Function->getName() << "':\n";
+
+    // start root signature header
+    Space++;
+    printSpaces(OS, Space);
+    OS << "Flags: " << format_hex(MRS.Flags, 8) << ":\n";
+    Space--;
```

Comment:
I think you can use `OS << indent(1) << format_hex(...` here instead of `printSpaces`